### PR TITLE
Legacy frontend integration test example 

### PR DESCRIPTION
**Pull request by tamarasaurus**
**Description (for Contributor and Core Developer)**

This PR is an example of how to load the product grid in the frontend integration tests, provide it with mock data, and run some assertions. We could use this setup to convert UI-only behats tests to save time on the CI.

To run the integration tests first build the PIM:

```bash
make php-image-dev
make dependencies
make pim-behat
```

Then you can execute 
```
yarn integration
```

To enable debug (which opens the browser window with devtools) mode you can run: 
```bash
DEBUG=true yarn integration
```

Note: For debug mode it should be run outside of Docker. 

We already have the PIM set up to run frontend integration tests with Jest and Puppeteer since 2.3. For the integration tests, Jest is configured in `tests/front/integration/jest`. Each file has a purpose: 

- **environment.js**  - Open a browser page when a test begins and set up the PIM to start
- **integration.jest.js** - A configuration file for Jest to know where the test files are and how to run them
- **setup.js** - Launch puppeteer and make the browser object accessible globally by the tests
- **teardown.js** - When a test finishes, shut down puppeteer and the browser

At the moment we only have one integration test for the family fetcher - 
https://github.com/akeneo/pim-community-dev/blob/3.2/tests/front/integration/pimenrich/fetcher/family-fetcher.integration.js

In this PR there are the following changes: 
- Add an integration test to set the value of a number filter on the product grid
- Add builders for mock data:
  - **DatagridProduct** - Builds a product to be returned for the product grid endpoint - `/datagrid/product-grid/load`
  - **NumberFilter** - Builds a number filter with choices to be returned in `/datagrid/product-grid/attributes-filters`
- Update the environment.js file of Jest to allow it to render the PIM correctly (ie. with styles, translations, locale)
- Add a set of tools to load the product grid in tests, and mock the JSON responses for products, filters, views, columns, category tree
- Add a set of [Jest matchers](https://jestjs.io/docs/en/expect) to make assertions on the product grid. For example to set a filter value, to check that the right products are displayed on the grid. 

In this PR there are also JSON files used for mocking the responses for the category tree and some other endpoints. For these endpoints it's not necessary to use builders for the moment. 

Puppeteer rendering the product grid with fake backend:
Screenshot from 2019-10-01 11-10-29

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | -
| Added legacy Behats               | -
| Added acceptance tests            | -
| Added integration tests           | -
| Changelog updated                 | -
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | -
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: tests/front/integration/jest/environment.js [hunk 1]
```diff
@@ -26,16 +29,67 @@ class PuppeteerEnvironment extends NodeEnvironment {
...
+        return req.respond({
           contentType: 'text/html;charset=UTF-8',
           body: baseFile,
         });
       }
+
+      if (req.url() === 'http://pim.com/js/extensions.json') {
+        return req.respond({
+          contentType: 'application/json',
+          body: JSON.stringify(require(`${process.cwd()}/public/js/extensions.json`))
```
Comment (tamarasaurus) [inline tests/front/integration/jest/environment.js]: Replace these with answerJson from the test tools

File: tests/front/integration/pimenrich/product-grid/grid-tools.js [hunk 2]
```diff
@@ -0,0 +1,162 @@
+const getOperatorChoiceByLabel = async (filter, choiceLabel) => {
```
Comment (tamarasaurus) [inline tests/front/integration/pimenrich/product-grid/grid-tools.js]: Add JSDOC

File: tests/front/integration/pimenrich/product-grid/grid-tools.js [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+  })
+
+  return { productGridData, productLoadData };
+}
+
+const buildProductGridResponses = (page, products = [], filters = []) => {
+  const { productGridData, productLoadData } = constructProductsResponse(products)
+
+  const mockResponses = {
+    'http://pim.com/datagrid_view/rest/product-grid/default-columns': ['application/json', require('./responses/default-columns.json')],
```
Comment (tamarasaurus) [inline tests/front/integration/pimenrich/product-grid/grid-tools.js]: Remove need to pass contentType

File: tests/front/integration/pimenrich/product-grid/grid-tools.js [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+      await (await filter.$('.AknFilterBox-filter')).click();
+
+      await page.waitForSelector(`${filterSelector} .filter-criteria`, {
+        visible: true,
+        timeout: 500
+      })
+
+      await (await filter.$('.operator')).click();
+      await (await getOperatorChoiceByLabel(filter, operator)).click();
+      await page.waitFor(500);
```
Comment (tamarasaurus) [inline tests/front/integration/pimenrich/product-grid/grid-tools.js]: Replace with waitForSelector on operator highlight

**Comment by juliensnz** [comment]
Really nice proposition. I have some doubt about how it will look if we do a fair amount of tests like that: will it still be usable and readable. It already seems pretty hard to follow.
Second, how do you sync the responses you have from the back and the mock you are doing here. With this current state, if the backend response fail, you will not be aware of it.
Last: I’m not a big fan of the waitFor(some time). It can make the test really prone to time related random fails

File: tests/front/integration/jest/environment.js [hunk 5]
```diff
@@ -26,16 +29,67 @@ class PuppeteerEnvironment extends NodeEnvironment {
...
+
+      if (req.url() === 'http://pim.com/rest/security/') {
+        return req.respond({
+          contentType: 'application/json',
+          body: JSON.stringify(require('../responses/rest-security.json'))
+        })
+      }
+
+      if (req.url() === 'http://pim.com/rest/user/') {
+        return req.respond({
```
Comment (arnolanglade) [inline tests/front/integration/jest/environment.js]: Why don't we use a lib like https://www.npmjs.com/package/fetch-mock ?

File: tests/front/common/builder/filters/number.js [hunk 6]
```diff
@@ -0,0 +1,96 @@
+/**
+ * Generate a datagrid number filter
+ *
+ * Example:
+ * const NumberFilterBuilder = require('../../common/builder/filter/number.js');
+ * const filter = (new NumberFilterBuilder()).withName('count').withLabel('Count').build();
+ */
+
+class NumberFilterBuilder {
```
Comment (arnolanglade) [inline tests/front/common/builder/filters/number.js]: I like the idea :)

File: tests/front/common/builder/filters/number.js [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+    }
+  }
+
+  withName(name) {
+    this.filter.name = name
+
+    return this;
+  }
+
+  withLabel(label) {
```
Comment (arnolanglade) [inline tests/front/common/builder/filters/number.js]: do we introduce mutator event if you don't need them so far? Dunno!

File: tests/front/integration/jest/environment.js (hunk 5 above)
Comment (pchasle) [reply tests/front/integration/jest/environment.js]: Mostly because `fetch-mock` will just handle the browser `fetch` API and we are not using this (e.g. jquery with xhr)
